module/zfs/zio.c

```diff
@@ -153,7 +153,8 @@ zio_init(void)
 		size_t size = (c + 1) << SPA_MINBLOCKSHIFT;
 		size_t p2 = size;
 		size_t align = 0;
-		size_t cflags = (size > zio_buf_debug_limit) ? KMC_NODEBUG : 0;
+		size_t cflags = (size > zio_buf_debug_limit) ?
+		    KMC_NODEBUG | KMC_KVMEM : KMC_KVMEM;
```

Comment:
Similarly, I don't think we need to change this code.  Instead what I'd suggest, is that we make the following small change to `spl_kmem_cache_create` so it defaults to `KMC_KVMEM` caches instead of `KMC_VMEM` caches for large objects.  This should always be a performance win for `kvmalloc()`.  Callers may still specifically request a `KMC_VMEM` cache.

```diff --git a/module/spl/spl-kmem-cache.c b/module/spl/spl-kmem-cache.c
index 44e112c..fa92a86 100644
--- a/module/spl/spl-kmem-cache.c
+++ b/module/spl/spl-kmem-cache.c
@@ -966,9 +966,9 @@ spl_kmem_cache_create(char *name, size_t size, size_t align,
                } else {
                        /*
                         * All other objects are considered large and are
-                        * placed on vmem backed slabs.
+                        * placed on kvmem backed slabs.
                         */
-                       skc->skc_flags |= KMC_VMEM;
+                       skc->skc_flags |= KMC_KVMEM;
                }
        }
```

Reply:
Aggreed